services/py-api/tests/unit_tests/test_service_layer/test_participant_registration_service.py

```diff
@@ -0,0 +1,172 @@
+
+import pytest
+from result import Ok, Err
+
+from src.database.model.participant_model import Participant
+from src.database.model.team_model import Team
+from src.server.exception import DuplicateTeamNameError, DuplicateEmailError, HackathonCapacityExceededError
+from src.server.schemas.request_schemas.schemas import ParticipantRequestBody
+from src.service.participants_registration_service import ParticipantRegistrationService
+
+
+@pytest.fixture
+def p_reg_service(hackathon_service_mock: Mock) -> ParticipantRegistrationService:
+    return ParticipantRegistrationService(hackathon_service_mock)
+
+
+@pytest.mark.asyncio
+async def test_register_admin_participant_success(
+    p_reg_service: ParticipantRegistrationService,
+    hackathon_service_mock: Mock,
+    team_repo_mock: Mock,
+    participant_repo_mock: Mock,
+    mock_input_data: ParticipantRequestBody,
+) -> None:
+    # Mock not full hackathon
+    hackathon_service_mock.check_capacity_register_admin_participant_case = AsyncMock(return_value=True)
+
+    # Mock successful `create` responses for team and participant. These are the operations inside the passed callback
+    # to with_transaction
+    team_repo_mock.create.return_value = Team(name=mock_input_data.team_name)
+    participant_repo_mock.create.return_value = Participant(
+        name=mock_input_data.name,
+        email=mock_input_data.email,
+        is_admin=True,
+        team_id=team_repo_mock.create.return_value.id,
+    )
+
+    hackathon_service_mock.create_participant_and_team_in_transaction.return_value = Ok(
+        (participant_repo_mock.create.return_value, team_repo_mock.create.return_value)
+    )
+
+    # Call the function under test
+    result = await p_reg_service.register_admin_participant(mock_input_data)
+
+    # Check that the result is an `Ok` containing both the participant and team objects
+    assert isinstance(result, Ok)
+    assert isinstance(result.ok_value, tuple)
+    assert isinstance(result.ok_value[0], Participant)
+    assert isinstance(result.ok_value[1], Team)
+
+
+@pytest.mark.asyncio
+async def test_register_admin_participant_duplicate_team_name_error(
+    p_reg_service: ParticipantRegistrationService,
+    hackathon_service_mock: Mock,
+    mock_input_data: ParticipantRequestBody,
+) -> None:
+    # Mock not full hackathon
+    hackathon_service_mock.check_capacity_register_admin_participant_case = AsyncMock(return_value=True)
+
+    # Mock `create_participant_and_team_in_transaction` to return an `Err` for duplicate team name
+    hackathon_service_mock.create_participant_and_team_in_transaction.return_value = Err(
+        DuplicateTeamNameError(mock_input_data.team_name)
+    )
+
+    # Call the function under test
+    result = await p_reg_service.register_admin_participant(mock_input_data)
+
+    # Check that the result is an `Err` with `DuplicateTeamNameError`
+    assert isinstance(result, Err)
+    assert isinstance(result.err_value, DuplicateTeamNameError)
```

Comment:
The comments are so on point. and I love the strategy that we are following for testing, now it actually makes sense. 🚀